### ISSUE TITLE
stabilized event called repeatedly - develop branch

### ISSUE THREAD

**Issue by mindade** [almende/vis#344]
I've integrated the code in #303 to handle dragging nodes. After the network is stabilized the stabilized event gets called repeatedly. 

As a side note, if I unfix all the nodes (i.e., return them to their normal state), the events stop. If I fix them, the events start again.

I think there might be some kind of recursive invocation going on here. Fixing the nodes may cause a start(), which eventually causes the stabilized event to fire, which causes me to fix the nodes again.

**Comment by AlexDM0** [comment]
I have talked to people with very big networks who wanted a stabilize event on EVERY stabilization to show a loading bar or something similar, some people want only the first one.. I could make two different events: initialStabilization and stabilization? What do you think?

**Comment by felixhayashi** [comment]
Hi @AlexDM0,

well, separating the events would allow us to remove the overhead of creating a flag that indicates if a network has already stabilized once, or not? This would be nice of course. Actually, I never understood why the stabilized event is called more than once. I mean stabilized means, it doesn't move on its own, right?

By the way, I noticed that for some reasons the threshold set via `stabilizationIterations` is sometimes exceeded (I will create a new ticket for that). I expected this parameter to force a stop.

-Felix

**Comment by AlexDM0** [comment]
Hi Felix,

Stabilization means it does not move. This can be initially when all the nodes start out randomly or when kick one around and wait for it to settle. The downside is that some functions call the start() method if the network is already settled, which will then immediately say stabilized! I'll see if we can avoid that specific case.

Now, as for the stabilizationIterations, this is used to "hide" the network and use full CPU power to attempt to stablize the network. This can be quite fast since we won't be tied to 1 physics tick per frame and we don't have to render anything. The downside here is that it freezes the network for a while. The stabilizationIterations denote how many physics ticks are allowed before starting to show the network on screen, not how many iterations are done before the network decides it is finished. The reason the event gives a number of iterations is that you can use that to determine how many you'd need (on average) if you wanted to do it beforehand.

Hope that clears it up.

I think that if we can avoid the firing of the stabilization when it is already stabilized, it should be sufficient.

Regards,

Alex

**Comment by felixhayashi** [comment]
Hi Alex,

Big thanks for explaining this!

> The stabilizationIterations denote how many physics ticks are allowed before starting to show the network on screen, not how many iterations are done before the network decides it is finished.

I always misunderstood this!

> I think that if we can avoid the firing of the stabilization when it is already stabilized, it should be sufficient.

I think so too. The dispatch of multiple stabilization events irritated me a lot in the beginning since stabilized means "stopped moving" and if the user didn't touch or modify the graph, how can the equilibrium be distorted and another "stabilized" be triggered...

Regards
Felix

**Comment by AlexDM0** [comment]
Hi Felix, 

The reason is that stabilized is not really stabilized. The timestep we take is a little big. This makes it fast (less iterations). We do a physics simulation, these simulations have equilibria and not all of them are stable (http://en.wikipedia.org/wiki/Mechanical_equilibrium) What we do now, is check if there is anything that is moving with a higher speed than our set speed, if that is not the case, we assume it is stable (enough). We simply stop the simulation. This does not mean all nodes would remain in that position if the simulation would continue. 

Now what happens is that when an option is changed, data is set or updated etc, we continue the simulation. I have tested with having to "not move" 5 times before we accept the solution, but even then can the next tick (which also gets 5 tries) move again. A solution to this is to 'revert' the state of 5 steps ago if it turns out that was stable. But imagine having a large dataset and all positions of the nodes and support nodes would be cached 5 times over... 

If I take a smaller timestep then yes, it becomes stable. But that also means we need more steps to reach our solution. To maximize speed, we try to make it JUST stable enough. Additionally there is a function in there that doubles the amount of steps per rendering if it is  a small network so everything is more responsive. All in all we want to avoid stuttering! A smooth experience is more important than a fast one.

From my tests, I think we cannot immediately discard secondary stabilization events. They sometimes take up to 40 or even 500 iterations (imagine a very slow moving small network, hardly visible but slow and constant). We could filter out the REALLY small ones, but that would only make this less predictable.

What's your opinion on this?

Regards,

Alex

**Comment by felixhayashi** [comment]
Hi Alex,

Yes, really difficult stuff.

> What we do now, is check if there is anything that is moving with a higher speed than our set speed, if that is not the case, we assume it is stable (enough). We simply stop the simulation. This does not mean all nodes would remain in that position if the simulation would continue. [...] Now what happens is that when an option is changed, data is set or updated etc, we continue the simulation.

Now I understand why occasionally a whole bunch of nodes starts to move again, even if a node insert occured far outside or if I changed the options.

>  But imagine having a large dataset and all positions of the nodes and support nodes would be cached 5 times over... 

Yes, I understand.

> If I take a smaller timestep then yes, it becomes stable. But that also means we need more steps to reach our solution. To maximize speed, we try to make it JUST stable enough.

My first shot was also to say, increase the precision so you can be sure that stable is really stable.

> A smooth experience is more important than a fast one.

I agree and increasing the number of calculations between the "rendering loops" goes against this.

> What's your opinion on this?

Maybe you could outsource the decision how the stabilization behaves to the api user somehow? For example I would prefer if the measuring precision is in sync with the actual calculation so I can be sure stabilized is fired once and really means stabilized. If the graph takes more than (half) a second, no problem, I could start displaying a loading bar saying "rendering please wait" - obviously this only occurs when calculating large graphs and the user will understand this. On the other hand an api user that really depends on speed could increase the calculation rate between the "game loop"...

-Felix

**Comment by AlexDM0** [comment]
I tend to disagree there. We're not doing scientific calculations. The solutions do not have to be accurate, they have to look good, be fast and be smooth. Having it mathematically stable is actually kind of irrelevant as long as the nodes have stopped moving, look good and the cpu can take some time off. 

The only real downside is that we have given the developers insight in a part of the engine that should not concern them. Maybe we should ignore secondary stabilization events alltogether. 

The configurePhysics menu was a solution to let people play with the physics engine. If we gave them numeric control over the timestep, instabilities would be so frequent that I'm afraid half the issues here would be about high cpu loads, never stable, long loading times, disappearing nodes (going to inf) etc. 

I think the best we can do is try to give the functionality everyone needs, without having them caring how.  

We could cache the nodes once and revert to that once we reach a stable point. As it is deterministic, it would always be stable. It's a bit of a memory eater but it could be done. There's memory enough on devices nowadays. 

Regards,

Alex

**Comment by felixhayashi** [comment]
Hi Alex,

> The configurePhysics menu was a solution to let people play with the physics engine. If we gave them numeric control over the timestep, instabilities would be so frequent that I'm afraid half the issues here would be about high cpu loads, never stable, long loading times, disappearing nodes (going to inf) etc. 

You are probably right with this.

> I think the best we can do is try to give the functionality everyone needs, without having them caring how. 

the quest for the holy grail :)

> We could cache the nodes once and revert to that once we reach a stable point. As it is deterministic, it would always be stable. It's a bit of a memory eater but it could be done. There's memory enough on devices nowadays.

This sounds good. I have not enough information to predict how well this works on the various devices but if you plan on using this "look ahead and rewind" technique I can surely report back how this works for people that use tiddlywiki.

-Felix

**Comment by felixhayashi** [comment]
Great, I will test it then :)
